### ISSUE TITLE
Good idea!

### ISSUE THREAD

**Issue by arabine** [marian-m12l/s9ke-toolchain#1]
Ah, this disassembly and emulator is a very good idea to see what's going on!
Your input file is the firmware located in the external flash memory?

**Comment by marian-m12l** [comment]
Thanks. I figured I'd look again for datasheets and was quite surprised to find the instruction set in the programming guide :-)
I went the easy and fast way with these python scripts, but I felt like I needed to at least make sure it would be somewhat reliable. Would be nice to make a proper Ghidra module for the disassembler at some point (contrary to most people in the RE community, I like that it's written in Java :-)).
I'm also starting a basic assembler, as I want to experiment with homebrew :-) Would also be handy to validate some assumptions, as the documentation is somewhat ambiguous or simply missing for some parts (would have be nicer to find the programming guide for the SNC7001A, as the memory map and MMIO do not match perfectly). USB in particular seems to be documented in another (unknown to me) document.

Would you happen to have looked at the breakout header? My device has a black blob on the microcontroller so i cannot check the pinout. All I know is that one of those seem to act like the reset pin when pulled to ground. Would be great if UART was broken out.

Oh, and yes, the input file is the whole content of the external chip. From what I understand, the DSP loads the first 32K words from the ROM chip into internal PRAM and runs from address `0x000000` (the reset interrupt handler).

**Comment by arabine** [comment]
Ok nice! On my side I have two objectives: dump the SPI memory and find a way to remove the black blob. I know who to ask for that.

**Comment by marian-m12l** [comment]
That black blob is a pain. Maybe you have already seen it, a user uploaded a few high res pictures of a PCB without this blob [here](https://github.com/marian-m12l/studio/discussions/145), why no blob on that unit, go figure... Also, there's big progress on the cipher thanks to a user reversing the ARM firmware :-)

**Comment by arabine** [comment]
Yes, you'll find more photos with the Sonix revealed on the Hackaday project. I whish I could get the ARM firmware I could help you more.
Will you continue this Sonix project now that we have the ARM firmware revealed?

**Comment by marian-m12l** [comment]
For sure the ARM tooling is much better than that of the DSP :-D
Regarding this project, I can't really tell yet. One big motivation was the reversing of the official firmware. Now that it seems out of the picture, this leaves mostly the possibility of developing homebrew firmware. I want to pursue this, but there's also many other things I want to try as well :-)
